### ISSUE TITLE
`<mdspan>`: `std::extents`'s CTAD unconditionally constrains argument types to be default-constructible

### ISSUE THREAD

**Issue by hewillk** [llvm/llvm-project#68671]
https://github.com/llvm/llvm-project/blob/f5031c655ddab72bdeccf5f830a17448926254e2/libcxx/include/__mdspan/extents.h#L457-L459

..which rejects the following code:

https://godbolt.org/z/47vfGne7s

```cpp
#include <mdspan>

struct S {
  S(size_t);
  operator size_t() const noexcept;
};

int main() {
  int* p;
  std::mdspan ms(p, S{3}); // ok
  std::extents exts(S{3}); // error
}
```

**Comment by ldionne** [comment]
@crtrott or @dalg24 Can you folks take a look?

**Comment by crtrott** [comment]
Yeah this is a bug, let me try to fix this.

**Comment by crtrott** [comment]
I believe changing the deduction guide implementation to this: 

```c++
template <class... _IndexTypes> 
extents(_IndexTypes...) -> extents<size_t, size_t((sizeof(_IndexTypes), dynamic_extent))...>;
```

should fix the problem. ANyt thoughts?

**Comment by ldionne** [comment]
Oh, I get it. We were using `_IndexTypes()` just as a way to get a variadic expansion but that was forcing `_IndexTypes` to be default constructible. Yeah your fix sounds good.

**Comment by crtrott** [comment]
Looks like I get warnings for left side of the operator not having any effect. This does not trigger warnings:

```c++
template <class... _IndexTypes> 
extents(_IndexTypes...) -> extents<size_t, size_t((sizeof(_IndexTypes)? dynamic_extent:dynamic_extent))...>;
```

Thoughts?

**Comment by crtrott** [comment]
Alternative would be introducing 
```c++
constexpr size_t just_dynamic_extent(size_t) { return dynamic_extent; }
```
helper function.

**Comment by ldionne** [comment]
This would be the usual way of dealing with the warning (cast to `void`):

```
template <class... _IndexTypes> 
extents(_IndexTypes...) -> extents<size_t, size_t(((void)sizeof(_IndexTypes), dynamic_extent))...>;
```

**Comment by crtrott** [comment]
ah ok I like that.

**Comment by crtrott** [comment]
Created a PR. @hewillk you want to check too that it does what you want? I added a test.

**Comment by crtrott** [comment]
And thanks for the report @hewillk. Another corner case we had not thought of.

**Comment by hewillk** [comment]
> And thanks for the report @hewillk. Another corner case we had not thought of.

Thanks for the quick fix.
In addition, there is a similar issue with the `extents`'s constructor, so let me report it directly here:
https://github.com/llvm/llvm-project/blob/f5031c655ddab72bdeccf5f830a17448926254e2/libcxx/include/__mdspan/extents.h#L313-L318

I believe we're missing `std::move` here:

https://godbolt.org/z/TrPqr5716

```cpp
#include <mdspan>

struct Index {
  operator size_t() && noexcept;
};

int main() {
  std::extents exts(Index{}); // error
}
```

**Comment by ldionne** [comment]
@hewillk Can you please open a new issue with that description? It helps us keep things tidy to treat only one problem per Github issue.
